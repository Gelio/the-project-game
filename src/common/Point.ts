export class Point {
  public readonly x: number;
  public readonly y: number;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  public static manhattanDistance(point1: Point, point2: Point): number {
    return Math.abs(point1.x - point2.x) + Math.abs(point1.y - point2.y);
  }
}
