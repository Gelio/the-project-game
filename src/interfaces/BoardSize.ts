export interface BoardSize {
  /**
   * Board width
   */
  x: number;

  /**
   * Task (neutral) area height
   */
  taskArea: number;

  /**
   * Each team's goal area height
   */
  goalArea: number;
}
