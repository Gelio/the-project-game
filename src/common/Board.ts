import { BoardSize } from '../interfaces/BoardSize';

export class Board {
  public readonly size: BoardSize;

  constructor(size: BoardSize) {
    this.size = size;
  }

  public reset() {
    // TODO: implement this
  }
}
