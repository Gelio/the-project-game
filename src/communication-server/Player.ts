import { LoggerInstance } from 'winston';

import { Communicator } from '../common/Communicator';

import { Message } from '../interfaces/Message';
import { Service } from '../interfaces/Service';

import { MessageRouter } from './MessageRouter';
import { PlayerInfo } from './PlayerInfo';

export class Player implements Service {
  public readonly info: PlayerInfo;

  private readonly communicator: Communicator;
  private readonly messageRouter: MessageRouter;
  private readonly logger: LoggerInstance;

  public get id() {
    return this.info.id;
  }

  constructor(
    communicator: Communicator,
    messageRouter: MessageRouter,
    logger: LoggerInstance,
    playerInfo: PlayerInfo
  ) {
    this.communicator = communicator;
    this.messageRouter = messageRouter;
    this.logger = logger;
    this.info = playerInfo;

    this.handleMessage = this.handleMessage.bind(this);
  }

  public init() {
    this.messageRouter.registerPlayerCommunicator(this.id, this.communicator);
    this.communicator.on('message', this.handleMessage);
  }

  public destroy() {
    this.logger.verbose(`Destroying player ${this.id}`);

    if (this.messageRouter.hasRegisteredPlayerCommunicator(this.id)) {
      this.messageRouter.unregisterPlayerCommunicator(this.id);
    }
  }

  private handleMessage<T>(message: Message<T>) {
    if (!this.isMessageValid(message)) {
      this.logger.warn(
        `Received message with sender ID ${message.senderId} but player ID is ${this.id}`
      );

      return;
    }

    this.messageRouter.sendMessageToGameMaster(this.info.gameName, message);
  }

  private isMessageValid<T>(message: Message<T>) {
    if (!this.id) {
      return true;
    }

    return this.id === message.senderId;
  }
}
