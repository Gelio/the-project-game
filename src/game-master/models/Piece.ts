import { Point } from '../../common/Point';

export class Piece {
  public isSham: boolean;

  /**
   * When the piece is picked up, this is the position of the player
   */
  public position: Point;
  public isPickedUp: boolean;
}
