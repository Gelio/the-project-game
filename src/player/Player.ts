import { createConnection } from 'net';
import { LoggerInstance } from 'winston';

import { config } from './config';

import { bindObjectProperties } from '../common/bindObjectProperties';
import { Communicator } from '../common/Communicator';
import { LoggerFactory } from '../common/logging/LoggerFactory';
import { UITransport } from '../common/logging/UITransport';
import { TeamId } from '../common/TeamId';

import { Message } from '../interfaces/Message';
import { PlayerAcceptedMessage } from '../interfaces/messages/PlayerAcceptedMessage';
import { PlayerHelloMessage } from '../interfaces/messages/PlayerHelloMessage';
import { PlayerRejectedMessage } from '../interfaces/messages/PlayerRejectedMessage';
import { MessageWithRecipient } from '../interfaces/MessageWithRecipient';
import { ListGamesRequest } from '../interfaces/requests/ListGamesRequest';
import { ListGamesResponse } from '../interfaces/responses/ListGamesResponse';
import { Service } from '../interfaces/Service';

import { registerUncaughtExceptionHandler } from '../registerUncaughtExceptionHandler';

import { UIController } from './ui/UIController';

export interface PlayerOptions {
  serverHostname: string;
  serverPort: number;
  askLevel: number;
  respondLevel: number;
  teamNumber: TeamId;
  teamLeader: boolean;
  timeout: number;
}

export class Player implements Service {
  private id: number;
  // @ts-ignore
  private isAccepted = false;

  private options: PlayerOptions;

  private communicator: Communicator;
  private logger: LoggerInstance;
  private readonly uiController: UIController;
  private readonly loggerFactory: LoggerFactory;

  private messageHandlers: { [type: string]: Function } = {
    PLAYER_ACCEPTED: this.handlePlayerAccepted,
    PLAYER_REJECTED: this.handlePlayerRejected
  };

  constructor(options: PlayerOptions, uiController: UIController, loggerFactory: LoggerFactory) {
    this.options = options;

    this.uiController = uiController;
    this.loggerFactory = loggerFactory;

    bindObjectProperties(this.messageHandlers, this);
    this.handleMessage = this.handleMessage.bind(this);
    this.handleServerDisconnection = this.handleServerDisconnection.bind(this);
  }

  public init() {
    const { serverHostname, serverPort } = this.options;

    this.initUI();
    this.initLogger();

    const socket = createConnection(
      {
        host: serverHostname,
        port: serverPort
      },
      () => {
        this.logger.info(`Connected to the server at ${serverHostname}:${serverPort}`);
        this.sendListGames();
        this.sendHandshake();
      }
    );

    this.communicator = new Communicator(socket, this.logger);
    this.communicator.bindListeners();

    this.communicator.on('message', this.handleMessage);
    this.communicator.on('close', this.handleServerDisconnection);
  }

  public destroy() {
    this.communicator.destroy();
    this.uiController.destroy();
  }

  private handleServerDisconnection() {
    this.destroy();
    this.logger.info('Disconnected from the server');
  }

  private sendHandshake() {
    const temporaryId = Math.floor(Math.random() * config.maxTemporaryPlayerId);

    const message: PlayerHelloMessage = {
      type: 'PLAYER_HELLO',
      senderId: -2,
      payload: {
        isLeader: this.options.teamLeader,
        teamId: this.options.teamNumber,
        temporaryId,
        game: 'the project game'
      }
    };

    this.communicator.sendMessage(message);
  }

  private async sendListGames() {
    const message: ListGamesRequest = {
      type: 'LIST_GAMES_REQUEST',
      senderId: -2,
      payload: undefined
    };

    this.communicator.sendMessage(message);
    const listGamesResponse = <ListGamesResponse>await this.communicator.waitForSpecificMessage(
      (msg: MessageWithRecipient<ListGamesResponse>) => msg.type === 'LIST_GAMES_RESPONSE'
    );

    this.handleListGamesResponse(listGamesResponse);
  }

  private handleMessage<T>(message: Message<T>) {
    // @ts-ignore
    if (this.messageHandlers[message.type] !== undefined) {
      return this.messageHandlers[message.type](message);
    }
  }

  private handleListGamesResponse(message: ListGamesResponse) {
    const games = message.payload.games;
    for (const game of games) {
      this.logger.info(`\n Game \"${game.name}\" : \"${game.description}\"`);
      this.logger.info(
        `\n boardSize: x:${game.boardSize.x} taskArea:${game.boardSize.taskArea} goalArea: ${
          game.boardSize.goalArea
        }`
      );
      this.logger.info(
        `\n Goal limit:${game.goalLimit} teamSizes: 1:${game.teamSizes[1]} 2:${game.teamSizes[2]}`
      );
      this.logger.info(`\n ${JSON.stringify(game.delays)}`);
    }
  }

  private handlePlayerAccepted(message: PlayerAcceptedMessage) {
    this.id = message.payload.assignedPlayerId;
    this.isAccepted = true;

    this.logger.info(`Player accepted with id ${this.id}`);
  }

  private handlePlayerRejected(message: PlayerRejectedMessage) {
    this.logger.error(`Player rejected. Reason: ${message.payload.reason}`);
    this.destroy();
  }

  private initUI() {
    this.uiController.init();
    this.uiController.render();
  }

  private initLogger() {
    const uiTransport = new UITransport(this.uiController.log.bind(this.uiController));
    this.logger = this.loggerFactory.createLogger([uiTransport]);
    registerUncaughtExceptionHandler(this.logger);

    this.logger.info('Logger initiated');
  }
}
