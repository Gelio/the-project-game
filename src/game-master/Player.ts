import { Point } from '../common/Point';
import { TeamId } from '../common/TeamId';
import { Piece } from './models/Piece';

export class Player {
  public isConnected: boolean;
  public teamId: TeamId;
  public isLeader: boolean;
  public isBusy = false;

  public playerId: number;
  public position: Point | null = null;
  public heldPiece: Piece | null = null;
}
