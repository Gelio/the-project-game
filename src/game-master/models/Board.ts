import { Point } from '../../common/Point';
import { BoardSize } from '../../interfaces/BoardSize';
import { Player } from '../Player';
import { Piece } from './Piece';
import { Tile } from './tiles/Tile';

export class Board {
  public readonly size: BoardSize;
  public readonly tiles: Tile[][];
  public readonly pieces: Piece[] = [];

  constructor(size: BoardSize, tiles: Tile[][]) {
    this.size = size;
    this.tiles = tiles;
  }

  public reset() {
    // TODO: implement this
  }

  public getTileAtPosition(position: Point) {
    const tileColumn = this.tiles[position.x];
    if (!tileColumn) {
      throw new Error('Invalid X coordinate');
    }

    const tile = tileColumn[position.y];
    if (!tile) {
      throw new Error('Invalid Y coordinate');
    }

    return tile;
  }

  public addPlayer(player: Player) {
    this.getTileAtPosition(player.position).player = player;
  }

  public removePlayer(point: Point) {
    this.tiles[point.x][point.y].player = null;
  }

  public movePlayer(player: Player, newPosition: Point) {
    const previousTile = this.getTileAtPosition(player.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.player !== player) {
      throw new Error('Old player position corrupted');
    }

    if (newTile.player) {
      throw new Error('Two players cannot stand on the same tile');
    }

    previousTile.player = null;
    newTile.player = player;
  }

  public addPiece(piece: Piece) {
    if (this.pieces.indexOf(piece) !== -1) {
      throw new Error('Piece already added');
    }

    const tile = this.getTileAtPosition(piece.position);
    if (tile.piece) {
      throw new Error('Piece already exists at that position');
    }

    tile.piece = piece;
    this.pieces.push(piece);
  }

  public removePiece(piece: Piece) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    this.pieces.splice(index, 1);

    if (!piece.isPickedUp) {
      const tile = this.getTileAtPosition(piece.position);
      tile.piece = null;
    }
  }

  public movePiece(piece: Piece, newPosition: Point) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    const previousTile = this.getTileAtPosition(piece.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.piece !== piece) {
      throw new Error('Old piece position corrupted');
    }

    if (newTile.piece) {
      throw new Error('Cannot move a piece on a tile which already has one');
    }

    previousTile.piece = null;
    newTile.piece = piece;
    piece.position = newPosition;
  }
}
