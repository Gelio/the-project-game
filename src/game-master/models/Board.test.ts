import { Point } from '../../common/Point';

import { BoardSize } from '../../interfaces/BoardSize';

import { Player } from '../Player';

import { Board } from './Board';
import { Piece } from './Piece';

describe('[GM] Board', () => {
  const boardSize: BoardSize = {
    x: 30,
    goalArea: 5,
    taskArea: 30
  };
  let board: Board;
  const pointsLimit: number = 15;

  beforeEach(() => {
    board = new Board(boardSize, pointsLimit);
  });

  describe('(piece)', () => {
    let piece: Piece;
    let piecePosition: Point;

    beforeEach(() => {
      piece = new Piece();
      piecePosition = new Point(1, boardSize.goalArea + 1);
      piece.isPickedUp = false;
      piece.position = piecePosition;
      piece.isSham = false;
      board.addPiece(piece);
    });

    describe('addPiece', () => {
      it('should add piece on the board', () => {
        expect(board.getTileAtPosition(piecePosition).piece).toBe(piece);
      });

      it('should throw an error when piece is added twice', () => {
        expect(board.addPiece.bind(board, piece)).toThrowError('Piece already added');
      });

      it('should throw an error when piece is added on position where another piece already exists', () => {
        const samePositionPiece = new Piece();
        samePositionPiece.position = piecePosition;

        expect(board.addPiece.bind(board, samePositionPiece)).toThrowError(
          'Piece already exists at that position'
        );
      });
    });

    describe('removePiece', () => {
      it('should remove piece from board', () => {
        board.removePiece(piece);

        expect(board.getTileAtPosition(piecePosition).piece).toBeNull();
      });

      it('should throw an error when piece is removed from the board where it does not exist', () => {
        board.removePiece(piece);

        expect(board.removePiece.bind(board, piece)).toThrowError('Piece was not on board');
      });

      it('should not remove piece if it is picked up', () => {
        piece.isPickedUp = true;
        board.removePiece(piece);

        expect(board.getTileAtPosition(piece.position).piece).toBe(piece);
      });
    });

    describe('movePiece', () => {
      it('should move piece to new position', () => {
        const newPiecePosition = new Point(piecePosition.x + 1, piecePosition.y);
        board.movePiece(piece, newPiecePosition);

        expect(board.getTileAtPosition(newPiecePosition).piece).toBe(piece);
      });

      it('should remove piece from the old position after moving it', () => {
        const newPiecePosition = new Point(piecePosition.x + 1, piecePosition.y);
        board.movePiece(piece, newPiecePosition);

        expect(board.getTileAtPosition(piecePosition).piece).toBeNull();
      });

      it('should throw an error when piece is moved to already taken position', () => {
        const samePositionPiece = new Piece();
        const newPiecePosition = new Point(0, 1);
        samePositionPiece.position = newPiecePosition;

        board.addPiece(samePositionPiece);

        expect(board.movePiece.bind(board, piece, newPiecePosition)).toThrowError(
          'Cannot move a piece on a tile which already has one'
        );
      });

      it('should throw an error when piece was not added on the board before', () => {
        const pieceNotAddedToGame = new Piece();
        const newPiecePosition = new Point(piecePosition.x + 1, piecePosition.y);
        pieceNotAddedToGame.position = newPiecePosition;

        expect(board.movePiece.bind(board, pieceNotAddedToGame, newPiecePosition)).toThrowError(
          'Piece has not been added to the game previously'
        );
      });

      it('should throw an error when piece position does not match the one on the board', () => {
        piece.position = new Point(piecePosition.x + 1, piecePosition.y);

        expect(board.movePiece.bind(board, piece, piecePosition)).toThrowError(
          'Old piece position corrupted'
        );
      });
    });

    it('should be removed and added at old position', () => {
      board.removePiece(piece);
      board.addPiece(piece);

      expect(board.getTileAtPosition(piecePosition).piece).toBe(piece);
    });
  });

  describe('(player)', () => {
    let player: Player;

    beforeEach(() => {
      player = new Player();
      player.isBusy = player.isConnected = player.isLeader = true;
      player.teamId = 1;
      player.playerId = 'player1';
    });

    describe('addPlayer', () => {
      it('should place player on the board', () => {
        board.addPlayer(player);

        expect(player.position).toBeDefined();
        expect(board.getTileAtPosition(<Point>player.position).player).toBe(player);
      });

      it('should throw an error when player was already added to the board', () => {
        board.addPlayer(player);
        const playerPosition = player.position;

        expect(board.addPlayer.bind(board, player)).toThrowError(
          'Player is already added on board'
        );
        expect(player.position).toBe(playerPosition);
      });
    });

    describe('movePlayer', () => {
      it('should move player to new position', () => {
        board.addPlayer(player);
        // FIXME: set `newPosition` based on player's position
        const newPosition = new Point(1, 2);

        board.movePlayer(player, newPosition);

        expect(player.position).toBe(newPosition);
      });

      it('should throw an error when player position does not match the one on the board', () => {
        board.addPlayer(player);
        // FIXME: set `newPosition` based on player's position
        const newPosition = new Point(2, 4);

        player.position = newPosition;

        expect(board.movePlayer.bind(board, player, newPosition)).toThrowError(
          'Old player position corrupted'
        );
      });

      it('should throw an error when player is not present on board', () => {
        const newPosition = new Point(2, 4);

        expect(board.movePlayer.bind(board, player, newPosition)).toThrowError(
          'Player position is null'
        );
      });

      it('should throw an error when player tries to step on tile with another player', () => {
        board.addPlayer(player);

        const playerTwo = new Player();
        board.addPlayer(playerTwo);

        expect(board.movePlayer.bind(board, player, playerTwo.position)).toThrowError(
          'Two players cannot stand on the same tile'
        );
      });
    });

    it('should be removed from the board', () => {
      board.addPlayer(player);
      board.removePlayer(player);

      expect(player.position).toBeNull();
    });

    it('should receive new position', () => {
      player.teamId = 2;
      board.addPlayer(player);
      board.setRandomPlayerPosition(player);

      expect(player.position).toBeDefined();
    });
  });

  describe('(tiles)', () => {
    it('should throw an error about invalid X coordinate', () => {
      const point: Point = new Point(30, 1);

      expect(board.getTileAtPosition.bind(board, point)).toThrowError('Invalid X coordinate');
    });

    it('should throw an error about invalid Y coordinate', () => {
      const point: Point = new Point(29, 40);

      expect(board.getTileAtPosition.bind(board, point)).toThrowError('Invalid Y coordinate');
    });

    it('should return the tile if position is valid', () => {
      const point: Point = new Point(0, 0);

      expect(board.getTileAtPosition.bind(board, point)).toBeDefined();
    });
  });
});
