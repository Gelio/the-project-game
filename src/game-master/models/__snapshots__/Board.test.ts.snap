// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GM] Board (piece) addPiece should throw an error when piece is added on position where another piece already exists 1`] = `"Piece already exists at that position"`;

exports[`[GM] Board (piece) addPiece should throw an error when piece is added twice 1`] = `"Piece already added"`;

exports[`[GM] Board (piece) movePiece should throw an error when piece is moved to already taken position 1`] = `"Cannot move a piece on a tile which already has one"`;

exports[`[GM] Board (piece) movePiece should throw an error when piece position does not match the one on the board 1`] = `"Old piece position corrupted"`;

exports[`[GM] Board (piece) movePiece should throw an error when piece was not added on the board before 1`] = `"Piece has not been added to the game previously"`;

exports[`[GM] Board (piece) removePiece should throw an error when piece is removed from the board where it does not exist 1`] = `"Piece was not on board"`;

exports[`[GM] Board (player) addPlayer should throw an error when player was already added to the board 1`] = `"Player is already added on board"`;

exports[`[GM] Board (player) movePlayer should throw an error when player is not present on board 1`] = `"Player position is null"`;

exports[`[GM] Board (player) movePlayer should throw an error when player position does not match the one on the board 1`] = `"Old player position corrupted"`;

exports[`[GM] Board (player) movePlayer should throw an error when player tries to step on tile with another player 1`] = `"Two players cannot stand on the same tile"`;

exports[`[GM] Board (tiles) should throw an error about invalid X coordinate 1`] = `"Invalid X coordinate"`;

exports[`[GM] Board (tiles) should throw an error about invalid Y coordinate 1`] = `"Invalid Y coordinate"`;

exports[`[GM] Board setRandomPlayerPosition should throw when there is no position for new player 1`] = `"No free position for player"`;
